R/epidist.R

```diff
@@ -1,242 +1,609 @@
+    density(x, at = day_range),
     ylab = "",
     xlab = "Time since infection",
     type = "p",
     pch = 16,
-    main = "Probability Density Function"
+    main = main,
+    ...
   )
 
   # plot CDF
   plot(
     day_range,
-    x$cdf(day_range),
+    cdf(x, q = day_range),
     ylab = "",
     xlab = "Time since infection",
     type = "p",
     pch = 16,
     ylim = c(0, 1),
-    main = "Cumulative Distribution Function"
+    main = "Cumulative Distribution Function",
+    ...
   )
 
-  # add a plot title
-  graphics::title("Distributions", outer = TRUE, line = -1)
+  if (!is.null(title)) {
+    if (grepl(pattern = "intrinsic", x = title, ignore.case = TRUE)) {
+      line <- -1
+    } else {
+      line <- -15
+    }
+    # add a plot title
+    graphics::title(title, outer = TRUE, line = line)
+  }
+}
+
+#' PDF, CDF, PMF, quantiles and random number generation for `epidist` and
+#' `vb_epidist` objects
+#'
+#' @description The epidist object holds a probability distribution which can
+#' either be a continuous or discrete distribution. These are the density,
+#' cumulative distribution, quantile and random number generation functions.
+#' These operate on any distribution that can be included in an epidist object.
+#'
+#' @param x An `epidist` or `vb_epidist` object
+#' @param at The quantiles to evaluate at
+#' @param q The quantiles to evaluate at
+#' @param p The probabilities to evaluate at
+#' @param times The number of random samples
+#' @param ... further arguments passed to or from other methods
+#'
+#' @return If an `epidist` object is given a numeric vector is returned, if an
+#' `vb_epidist` object is given a list of two elements each with a numeric
+#' vector is returned
+#'
+#' @name epidist_distribution_functions
+NULL
+
+#' @rdname epidist_distribution_functions
+#' @export
+density <- function(x, at) UseMethod("density")
+
+#' @export
+density.epidist <- function(x, at) {
+  unlist <- length(x$prob_dist) == 1
+  if (inherits(x$prob_dist, "distcrete")) {
+    out <- x$prob_dist$d(at)
+  } else {
+    out <- stats::density(x$prob_dist, at = at)
+  }
+  out <- if (unlist) unlist(out, recursive = FALSE) else out
+  out
+}
+
+#' @rdname epidist_distribution_functions
+#' @export
+cdf <- function(x, q) UseMethod("cdf")
+
+#' @export
+cdf.epidist <- function(x, q) {
+  unlist <- length(x$prob_dist) == 1
+  if (inherits(x$prob_dist, "distcrete")) {
+    out <- x$prob_dist$p(q)
+  } else {
+    out <- distributional::cdf(x$prob_dist, q = q)
+  }
+  out <- if (unlist) unlist(out, recursive = FALSE) else out
+  out
+}
+
+#' @rdname epidist_distribution_functions
+#' @export
+quantile <- function(x, p) UseMethod("quantile")
+
+#' @export
+quantile.epidist <- function(x, p) {
+  unlist <- length(x$prob_dist) == 1
+  if (inherits(x$prob_dist, "distcrete")) {
+    out <- x$prob_dist$q(p)
+  } else {
+    out <- stats::quantile(x$prob_dist, p = p)
+  }
+  out <- if (unlist) unlist(out, recursive = FALSE) else out
+  out
+}
+
```

Comment:
Suggest adding examples to the functions grouped under this heading, including `density`, `cdf` etc; especially adding separate examples for the `epidist` and `vb_epidist` classes.

Reply:
Resolved in 8eb31ecde65b28f7e52e89004e3088aa3a90bf13